### PR TITLE
Optimize block date retrieval for course outlines.

### PR DESCRIPTION
**Pull request by doctoryes**
**Description:**

When a course outline is requested, currently *all* block dates in the course are returned. This PR adds an optional optimization that allows only the dates for outline-relevant block types to be returned (course, chapter, sequential), which should greatly improve response times.

Also, caching is re-worked to use the TieredCache which caches data in both the ephemeral request cache as well as Django's long-term cache.

I've also created an edx-platform PR which uses this edx-when version for testing and proving out these changes - it's here:
https://github.com/edx/edx-platform/pull/28551

**JIRA:** https://openedx.atlassian.net/browse/TNL-8061

**Dependencies:** None

**Merge deadline:** None

**Installation instructions:** None

**Testing instructions:**

The change has been deployed to this sandbox:
https://edx-when-juliasq.sandbox.edx.org/
in order to test normal operation.

**Reviewers:**
- [ ] @ormsbee 
- [ ] @mikix / @Dillon-Dumesnil 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** I've added notes throughout.

### CODE REVIEW

File: edx_when/api.py
Comment (Dillon-Dumesnil) [reply edx_when/api.py:233]: Do we have a condition on which we can remove the None? Or just always keep because we won't ever be able to guarantee everything has it filled?

File: edx_when/models.py
Comment (Dillon-Dumesnil) [inline edx_when/models.py:116]: Nit: Add course_id and location here using the Index format instead of db_index on the field definition.

File: edx_when/api.py
Comment (Dillon-Dumesnil) [inline edx_when/api.py:367]: Should this method also be clearing the cache when a new date is set? I know we do it when we clear dates, but wouldn't this be a similar process?

File: edx_when/models.py
Comment (Dillon-Dumesnil) [inline edx_when/models.py:87]: Why have this as its own model rather than just a field on a ContentDate?

**Review by Dillon-Dumesnil — COMMENTED** [review]
So I did my best to understand this, but I'm honestly not particularly well versed in caching so feel free to take all my comments with a grain of salt. I think it would be good to wait (if possible) for Mike Terry to return from vacation so he can also take a look. He is back Thursday

File: edx_when/api.py
Comment (ormsbee) [reply edx_when/api.py:127]: Does it need to be called though? If the version information is encoded into the cache key?

Comment (doctoryes) [reply edx_when/api.py:127]: I would have just removed it - but the fact that a `None` published_version can still be sent in made me think that the code still should clear the cache in that case. (I also hadn't made the edx-platform PR yet which uses the new edx-when, so I wasn't sure at the time if a published course version could be passed-in for all calls. I've since found out that a course version *can* be passed-in everywhere.)

Also, in the edx-platform PR, I'm getting the published version from `course.course_version` in most places. And I don't understand completely when I can be assured that attribute will be present and filled-in. In Old Mongo courses, wouldn't that attribute be `None` or non-existent?

File: edx_when/api.py
Comment (doctoryes) [reply edx_when/api.py:233]: This PR adds the new column with `null=True` to avoid DB locking on deployment. And I've chosen to do a lazy load of the column upon course publish. With this strategy, it's possible that some courses will never be published and won't have values for block_type ever - unless we populate the value for all courses using some asynchronous process. Once a course *is* published, no `None` values would be expected for block_type.

Summary: With the current plan, just always keep.

File: edx_when/api.py
Comment (doctoryes) [reply edx_when/api.py:367]: The assumption I'd made here is that a `set_date_for_block` would be preceded by a course publish - and therefore a new `published_version` would be passed-in. This new `published_version` would leave the old cached value behind. Clearing all course dates on blocks might *not* necessarily be preceded by a course publish.

File: edx_when/api.py
Comment (doctoryes) [inline edx_when/api.py:369]: FYI: while the `published_version` is saved on each `ContentDate` object, it's not used as a filter when querying course dates (though it *is* returned for each block).

File: edx_when/models.py
Comment (doctoryes) [reply edx_when/models.py:87]: Good question! Because I just can't help myself from normalizing DB data? 😸 It would also work fine as a field on `ContentDate` - I arbitrarily chose to use a new model.

File: edx_when/models.py
Comment (doctoryes) [reply edx_when/models.py:116]: Interesting - I'll try this and see how it changes the migration file.

**Review by doctoryes — COMMENTED** [review]
Thanks for the review, @Dillon-Dumesnil ! I'll discuss with my team about waiting for a merge. We'd like to get a PR approval from an owner of this repo ideally before proceeding.

File: edx_when/models.py
Comment (doctoryes) [inline edx_when/models.py:110]: @ormsbee In thinking more about this field, does it need to be present at all? The code currently does *not* use it to filter against. It's stored for `ContentDate` rows but never used to filter against in the current code - so it's really there just as a piece of info that might be useful for future debugging. The course_version/published_version is currently only used to invalidate old cache keys upon a new course publish.

Comment (ormsbee) [reply edx_when/models.py:110]: It sounds fine to remove.

File: edx_when/migrations/0008_courseversion_block_type.py
Comment (doctoryes) [inline edx_when/migrations/0008_courseversion_block_type.py:41]: Here's the SQL which this migration will run - generated with `./manage.py cms sqlmigrate edx_when 0008`:
```
BEGIN;
--
-- Add field block_type to contentdate
--
ALTER TABLE `edx_when_contentdate` ADD COLUMN `block_type` varchar(255) NULL;
--
-- Alter field course_id on contentdate
--
DROP INDEX `edx_when_contentdate_course_id_e6c39fdc` ON `edx_when_contentdate`;
--
-- Alter field location on contentdate
--
DROP INDEX `edx_when_contentdate_location_485206ea` ON `edx_when_contentdate`;
--
-- Create index edx_when_course_block_type_idx on field(s) course_id, block_type of model contentdate
--
CREATE INDEX `edx_when_course_block_type_idx` ON `edx_when_contentdate` (`course_id`, `block_type`);
--
-- Create index edx_when_course_id_idx on field(s) course_id of model contentdate
--
CREATE INDEX `edx_when_course_id_idx` ON `edx_when_contentdate` (`course_id`);
--
-- Create index edx_when_location_idx on field(s) location of model contentdate
--
CREATE INDEX `edx_when_location_idx` ON `edx_when_contentdate` (`location`);
COMMIT;
```

**Comment by doctoryes** [comment]
@Dillon-Dumesnil To clarify the concerns about approving this PR now:
- Is @mikix 's approval a pre-requisite for merging?
- Would your team feel comfortable with us merging/deploying this change based on our own review - and having @mikix review the changes post-deploy?
  - Naturally, our team would assume responsibility for the deployment and any needed fixes/rollback.
Thanks!

File: edx_when/api.py
Comment (Dillon-Dumesnil) [inline edx_when/api.py:397]: Thank you for these awesome comments! Always good to have some fresh eyes

File: edx_when/api.py
Comment (Dillon-Dumesnil) [inline edx_when/api.py:382]: Just me wondering, and no obligation at all to test or try or change anything, but I wonder why do the list and [0] vs just using `.first()` on the query set. Not sure the error it throws if none (maybe none) but maybe an option. Not at all for you to do, just commenting while in my head.

File: edx_when/api.py
Comment (Dillon-Dumesnil) [inline edx_when/api.py:404]: Super Nit: Put the comment here and on line 419 above the if/else lines (one line higher) rather than just below. I personally always interpret python comments on what is directly beneath so seeing this comment and then the UserDate below, I was like yeah, cool, checks out. But then on line 419, the if statement didn't click with the comment and I realized it was applying to the whole else block. It's a small thing, feel free to ignore

File: edx_when/api.py
Comment (Dillon-Dumesnil) [inline edx_when/api.py:380]: Thanks for DRYing it up in here

File: edx_when/migrations/0008_courseversion_block_type.py
Comment (Dillon-Dumesnil) [reply edx_when/migrations/0008_courseversion_block_type.py:41]: Ah whoops. I didn't mean for additional SQL to be executed via the switch to models.Index vs the db_index parameter. Happy to have you keep it, but if you'd rather not keep in these additional changes, totally understand.

**Review by Dillon-Dumesnil — APPROVED** [review]
Sorry for all the nits late in the game, but hey, they're nits. Do with them what you will. Biggest question is probably regarding the functionality (in the comment about subsection_and_higher_only variable name), but trust y'all's decision on that. 
Thanks for this PR and all of the little improvements you made throughout the repo. We always appreciate a PR that leaves the place better than it found it and this certainly hits that!

File: edx_when/api.py
Comment (doctoryes) [reply edx_when/api.py:28]: I'm not in 😍 with the name `subsection_or_higher_only` either - but it's the best I could think of and it's at least explicit without mentioning outlines. I agree with your proposed ideal solution as well. But I pulled back from that solution (passing in a list of block types) to avoid the extra scope & testing - and KISS/YAGNI.

File: edx_when/api.py
Comment (doctoryes) [reply edx_when/api.py:43]: Thanks for pointing that out! I'll change this to stay consistent.

File: edx_when/api.py
Comment (doctoryes) [reply edx_when/api.py:41]: Agree - I'll change it.

File: edx_when/api.py
Comment (doctoryes) [reply edx_when/api.py:153]: I'll remove - states the obvious.

File: edx_when/api.py
Comment (doctoryes) [reply edx_when/api.py:176]: Will do.

File: edx_when/api.py
Comment (doctoryes) [reply edx_when/api.py:382]: According to https://docs.djangoproject.com/en/2.2/ref/models/querysets/#first, using `first()` would work just as well. I simply copied the existing code, removing the never-thrown exception.

File: edx_when/api.py
Comment (doctoryes) [reply edx_when/api.py:397]: It took awhile to figure it out! So thought I'd capture it.

File: edx_when/api.py
Comment (doctoryes) [reply edx_when/api.py:404]: Will do.

File: edx_when/migrations/0008_courseversion_block_type.py
Comment (doctoryes) [reply edx_when/migrations/0008_courseversion_block_type.py:41]: I'm actually going to remove it - just to avoid risk by doing less index creation and dropping for *this* PR. Thanks!

File: edx_when/api.py
Comment (doctoryes) [reply edx_when/api.py:140]: Will do here and elsewhere in the file.

**Review by doctoryes — COMMENTED** [review]
Thanks so much for the re-review, @Dillon-Dumesnil ! I've addressed all your comments and will incorporate most of them.

File: edx_when/migrations/0008_courseversion_block_type.py
Comment (doctoryes) [reply edx_when/migrations/0008_courseversion_block_type.py:22]: I've removed the index changes as mentioned - the SQL that will be executed is below:
```
BEGIN;
--
-- Add field block_type to contentdate
--
ALTER TABLE `edx_when_contentdate` ADD COLUMN `block_type` varchar(255) NULL;
--
-- Create index edx_when_course_block_type_idx on field(s) course_id, block_type of model contentdate
--
CREATE INDEX `edx_when_course_block_type_idx` ON `edx_when_contentdate` (`course_id`, `block_type`);
COMMIT;
```

File: edx_when/api.py
Comment (Dillon-Dumesnil) [reply edx_when/api.py:28]: Yeah, that's fair. Thank you!

File: edx_when/__init__.py
Comment (mikix) [inline edx_when/__init__.py:2]: :chefkiss:

**Review by mikix — COMMENTED** [review]
Just a quick post-merge comment to say this looks good and thank you Julia for your work!

**Pull request merged**
